kernel/src/engine/default/filesystem.rs

```diff
@@ -72,7 +80,14 @@ impl<E: TaskExecutor> FileSystemClient for ObjectStoreFileSystemClient<E> {
             }
         });
 
-        Ok(Box::new(receiver.into_iter()))
+        if self.is_local {
```

Comment:
it seems like we might need to just detect if it "isn't S3 non-directory bucket" and then keep expanding from there? Seems safer to do a bit of extra sorting for now and just remove it as we discover different object_stores can do sorting? (or better yet we do some upstream fixes to expose a new API to communicate sorting)

Reply:
This will be huge for building any analytical engines, not knowing whether to build trust in the iterators to return the sorted entries. Are there any conversations going on in upstream in this regard?

Reply:
Yep! I added some context here https://github.com/delta-incubator/delta-kernel-rs/pull/344#issuecomment-2406146857. But I think we want to guarantee ordered from _our_ listing.

Reply:
Yet another reason: `object_store` supports various [adapters](https://docs.rs/object_store/latest/object_store/#adapters) which would throw off name- or type-based checks, so any such check needs to fail-safe not fail-dangerous.